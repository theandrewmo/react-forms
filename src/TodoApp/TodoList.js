import NewTodoForm from "./NewTodoForm";
import { useState } from 'react';
import Todo from './Todo';
import { DragDropContext, Droppable, Draggable } from '@hello-pangea/dnd';


/** TodoList - renders the NewTodoForm and the list of Todo Components
 * 
 * Props:
 * 
 * State:  [all todos]
 * 
 */

const TodoList = ()=> {    

    /** set initial state to an empty array */
    const INITIAL_STATE = []
    const [todos, setTodos] = useState(INITIAL_STATE)

    const addTodo = (newTodo) => {
        setTodos([...todos, newTodo])
    }

    const removeTodo = (todoId) => {
        setTodos(todos.filter(todo => todo.taskId !== todoId))
    }

    const editTodo = (todoId, updatedTask) => {
        setTodos((prevTodos) => (
            prevTodos.map((todo) => {
                if (todo.taskId === todoId) {
                    return { ...todo, task: updatedTask };
                }
                return todo;
            })
        ));
    };

    const markAsCompleted = (todoId) => {
        setTodos((prevTodos)=> (
            prevTodos.map((todo) => {
                if (todo.taskId === todoId) {
                    return {...todo, isCompleted: true}
                }
                return todo;
            })
        ))
    }

    const markAsNotCompleted = (todoId) => {
        setTodos((prevTodos)=> (
            prevTodos.map((todo) => {
                if (todo.taskId === todoId) {
                    return {...todo, isCompleted: false}
                }
                return todo;
            })
        ))
    }

    const onDragEnd = (result) => {
        if (!result.destination) {
          return;
        }
    
        const updatedTodos = [...todos];
        const [reorderedTodo] = updatedTodos.splice(result.source.index, 1);
        updatedTodos.splice(result.destination.index, 0, reorderedTodo);
    
        setTodos(updatedTodos);
      };
    
      return (
        <div className="max-w-md mx-auto mt-8">
          <NewTodoForm addTodo={addTodo} />
          <DragDropContext onDragEnd={onDragEnd}>
            <Droppable droppableId="todos" direction="vertical">
              {(provided) => (
                <div
                  ref={provided.innerRef}
                  {...provided.droppableProps}
                >
                  {todos.map((todo, index) => (
                    <Draggable key={todo.taskId} draggableId={todo.taskId} index={index}>
                      {(provided) => (
                        <div
                          {...provided.draggableProps}
                          {...provided.dragHandleProps}
                          ref={provided.innerRef}
                        >
                          <Todo
                            task={todo.task}
                            isCompleted={todo.isCompleted}
                            removeTodo={() => removeTodo(todo.taskId)}
                            editTodo={(updatedTask) => editTodo(todo.taskId, updatedTask)}
                            markAsCompleted={() => markAsCompleted(todo.taskId)}
                            markAsNotCompleted={() => markAsNotCompleted(todo.taskId)}
                          />
                        </div>
                      )}
                    </Draggable>
                  ))}
                  {provided.placeholder}
                </div>
              )}
            </Droppable>
          </DragDropContext>
        </div>
      );
}

export default TodoList;