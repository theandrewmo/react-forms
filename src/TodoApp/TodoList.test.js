import { render, screen, fireEvent } from '@testing-library/react';
import TodoList from './TodoList';

it('renders', () => {
  render(<TodoList />);
});

it("matches snapshot", () => {
    const { asFragment } = render(<TodoList />);
    expect(asFragment()).toMatchSnapshot();
})

it("adds a new todo", ()=> {
  render(<TodoList />)

  const taskInput = screen.getByTestId('task')
  const submitButton = screen.getByTestId('submitButton');

  fireEvent.change(taskInput, {target : {value : 'testTask'}})
  fireEvent.click(submitButton)

  expect(screen.getByTestId('todo')).toBeInTheDocument();
})