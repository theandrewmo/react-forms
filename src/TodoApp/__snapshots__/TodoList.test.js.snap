// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="max-w-md mx-auto mt-8"
  >
    <form
      class="mb-4"
    >
      <label
        class="block font-semibold mb-2"
        for="newTask"
      >
        New Task:
      </label>
      <input
        class="w-full p-2 border rounded mb-2 transition duration-300 transform hover:scale-105"
        data-testid="task"
        id="task"
        name="task"
        placeholder="what to do? e.g. do laundry"
        value=""
      />
      <button
        class="px-4 py-2 bg-blue-500 text-white rounded transition duration-300 transform hover:scale-110"
        data-testid="submitButton"
      >
        Add
      </button>
    </form>
    <div
      data-rbd-droppable-context-id="1"
      data-rbd-droppable-id="todos"
    />
  </div>
</DocumentFragment>
`;
